src/util/log.cc

```diff
@@ -542,28 +542,20 @@ std::ostream& operator<<(std::ostream& out, const std::exception_ptr& eptr) {
         // Print more information on some familiar exception types
         try {
             throw;
-        } catch (const seastar::nested_exception& ne) {
-            out << fmt::format(": {} (while cleaning up after {})", ne.inner, ne.outer);
-        } catch(const std::system_error &e) {
+        } catch (const std::system_error &e) {
             out << " (error " << e.code() << ", " << e.what() << ")";
-        } catch(const std::exception& e) {
+        } catch (const std::exception &e) {
             out << " (" << e.what() << ")";
-            try {
-                std::rethrow_if_nested(e);
-            } catch (...) {
-                out << ": " << std::current_exception();
-            }
-        } catch (const std::nested_exception& ne) {
-            // if we got here, it means that the nested_exception is not mixed in
-            // with an object of type std::exception. It can happen in one of two
-            // cases:
-            // 1. The std::nested_exception was built and thrown using a normal
-            //    throw and not in the context of another catch with std::throw_with_nested
-            // 2. Somewhere in the code there is a throw of a type which doesn't derive from
-            //    std::exception and then it is rethrown using std::throw_with_nested.
-            out << ": " << "unknown nested type: " << ne.nested_ptr();
-        } catch(...) {
-            out << " (" << "no extra info" << ")";
+        } catch (...) {
+            // no extra info
+        }
+
+        try {
+            throw;
```

Comment:
Why does this need to be a second try/throw - can't it just be in a catch clause in the try above? These throws are not free.

Reply:
@nyh I think I get it, this method eliminates two bugs - 
1. Is the one we already know - i.e std::nested_exception which is not derived from std::exception
2. A more significant one which @vladzcloudius  found - since before this change the `std::rethrow_if_nested` is placed inside the catch block for `std::exception`, the chain will break if one of the nested exceptions happens to be of type `std::system_error` , this is what vlad demonstrats in the example, the choices are either:
2.1. Do what vlad did (which IMHO is more generic)
2.2. Add an internal  `rethrow_if_nested` to every catch block and **in addition** have a separate catch block for std::nested_exception.

Reply:
You're right! Good catch about the nested_exception *also* being an std::system_error.
It's sad we always need to add a second slow throw just for this reason.

By the way, this issue reaffirms my belief that std::nested_exception sucks (at least for Seastar) and we should never have used it. The right approach (at least for Seastar) would have been to have an exception type which is a vector of exception pointers, but  @avikivity and others didn't like this proposal (because it's not C++-like because it isn't useful in a try/catch). Nice that despite those arguments, Rafael nevertheless added such a type - he just cleverly named seastar::nested_exception so nobody would notice it's different :-)

Anyway, whether or not we recommend seastar::nested_exception or use it exclusively in Seastar, I agree that a useful `operator<<` should support this standard type in a good way, because Seastar applications are fairly likely to be using this standard feature - unless we strongly advise against this type (and we don't, currently).

Reply:
Thanks, @eliransin.
@nyh, what Eliran has explained is also described in the patch comment: "In particular if std::system_error appears in the middle on the nested exception chain the logging will break and very important information (e.g. the object that initiated the exception chain) is going to be lost."

> can't it just be in a catch clause in the try above? These throws are not free.
My assumption is that this code is only hit on rare occasions when real errors occur. In this case we would care much less about the performance of the printing code.

And the robustness of this code is of the utter importance! And my patch is *mainly* about that - see more below.

Eliran described the alternative exactly the same way I envisioned it too: patching every "catch" with "rethrow_if_nested" and this is SUPER ugly to say the least. More importantly, this kind of codding is very likely going to cause this function to get broken again next time it's going to be touched.

So, how this may be improved?

While you were very clear about hating the idea of us handling special cases of particular exception type (printing specific object's info depending on its type or the type it derived from) - this is exactly what this code does today. 

We only handle 4 special cases so far: 
 * std::system_error
 * std::exception
 * seastar::nested_exception
 * std::nested_exception

In addition to that the same function (operator<<) unrolls the std::nested_exception using the recursion.

Therefore I thought that it makes sense to clearly separate these two things:
* Printing part
* Recursion part

This way, if we want to add more special cases to the "printing part" - it's, first, trivial, and, second, not going to change (break) the "recursion part".

Reply:
> @nyh, what Eliran has explained is also described in the patch comment: "In particular if std::system_error appears in the middle on the nested exception chain the logging will break and very important information (e.g. the object that initiated the exception chain) is going to be lost."

It wasn't clear to me that `std::system_error` here wasn't just an example, but rather discussing this specific type.

> > My assumption is that this code is only hit on rare occasions when real errors occur. In this case we would care much less about the performance of the printing code.

You're right that it's just about logging errors, which shouldn't be very common - otherwise the log size will explode. I still don't like to see unnecessary waste of time, but it is beginning to seem like this one is indeed necessary.
> While you were very clear about hating the idea of us handling special cases of particular exception type (printing specific object's info depending on its type or the type it derived from) - this is exactly what this code does today.
> 
> We only handle 4 special cases so far:
> 
>     * std::system_error

I think this was a mistake. The whole point of std::exception having having a `what()` is for it to be able to describe itself nicely. Isn't `std::system_error`'s `what()` good enough? I don't know who added this case, I hope it wasn't me ;-) 

> 
>     * std::exception

This is for the what(), and ideally should be the only one we support, but

> 
>     * seastar::nested_exception
> 
>     * std::nested_exception

At some point, Seastar stopped using std::nested_exception and switched to seastar::nested_exception. When we did that, we should have added a what() for seastar::nested_exception, which would have made this case unneeded either (with the small cost of extra allocations of the concatenated strings).

We didn't have a `std::nested_exception` case until Eliran's fix, and we don't really need it for Seastar. It is dubious why it's needed for Seastar-based applications, but maybe it doesn't hurt much.

Reply:
> > We only handle 4 special cases so far:
> > ```
> > * std::system_error
> > ```
> 
> I think this was a mistake. The whole point of std::exception having having a `what()` is for it to be able to describe itself nicely. Isn't `std::system_error`'s `what()` good enough? I don't know who added this case, I hope it wasn't me ;-)

Well, I'm not arguing that ideally we'd only need to print `exception::what()` overload output but by looking at the code above it seems that the system error code is (for some reason) not included in the `what()` for system_error. Hence the special casing.

> > ```
> > * seastar::nested_exception
> > 
> > * std::nested_exception
> > ```
> 
> At some point, Seastar stopped using std::nested_exception and switched to seastar::nested_exception. When we did that, we should have added a what() for seastar::nested_exception, which would have made this case unneeded either (with the small cost of extra allocations of the concatenated strings).

That would do, yes.
But this definitely deserves a separate patch. ;)

> 
> We didn't have a `std::nested_exception` case until Eliran's fix, and we don't really need it for Seastar. It is dubious why it's needed for Seastar-based applications, but maybe it doesn't hurt much.

Reply:
> You're right! Good catch about the nested_exception _also_ being an std::system_error.
> It's sad we always need to add a second slow throw just for this reason.

No, the second throw is needed not only because of that, @nyh.
The problem here is that std::nested_exception can also (at least potentially) include seastar::nested_exception in the chain too (or anything that can be added to a "special case list" in the future).

So, we should either instrument every `catch` of the "printing block" similarly to the std::exception one (super UGLY) or handle the recursion separately, like in the patch.